test(agw): Add scaling tests for DHCP ip allocation mode

MoritzThomasHuebner opened the pull request:
We had to revert #14724 in #14881 because of out of memory failures on the `magma_deb` VM. This issue was not apparent on the magma dev VM because mobilityd was able to write into the swap. Each `dhcp_helper_cli` thread spawned when we release one IP uses about 50 MB, so running six or more in parallel will cause an out of memory error (The default memory limit of mobilityd is 300 MB).

## Summary

- Revert #14881
- Add a mechanism to limit the number of concurrent dhcp helper cli threads to 3 by implementing a queue.
- Add a test utility to check that the correct number of leases are allocated and all leases are released at the end.
- Add a test utility to clear the lease file at the end of the test.
- Some minor refactoring to improve the traffic server utilities.

## Test Plan

We manually observed the memory usage and number of dhcp_helper_cli threads during test execution manually with `watch -n 0.1 "systemctl status magma@mobilityd"` and found that the 300MB limit was no longer exceeded.

MoritzThomasHuebner committed:
revert" "revert: "test(agw): Add scaling tests for dhcp mode (#14724)" (#14881)"

This reverts commit c86e4a9b1fd939a8ec7dc465f6cc7c965a20b506.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
feat(agw): Added limiting for dhcp helper cli processes

- Limit the number of dhcp helper cli to 3
- Implement a waiting queue for new cli threads

Co-authored-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -50,6 +50,10 @@
 class IPAllocatorDHCP(IPAllocator):
+
+    MAX_DHCP_PROCS = 3
+    dhcp_helper_procs = []
```
Good dog

nstng commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -435,30 +439,22 @@ def _release_dhcp_ip(self, ip_desc: IPDesc) -> None:
-                call_args,
-                capture_output=True,
-                check=False,
+            if len(self.dhcp_helper_procs) == self.MAX_DHCP_PROCS:
```
Is this thread safe? By a race condition, can we get in a state where len(self.dhcp_helper_procs) > self.MAX_DHCP_PROCS, i.e., the list will not be inspected any more? Might >= be more reasonable to mitigate this?

nstng replied:
If there are no _release_dhcp_ip calls for some time MAX_DHCP_PROCS=3 potentially stuck processes can be kept for a long time - should this be prevented?

MoritzThomasHuebner replied:
I made this a bit safer by changing it to `while len(self.dhcp_helper_procs) >= self.MAX_DHCP_PROCS`. That way we should safely remove any excess processes every time we want to release an IP. I believe the logic would be thread-safe regardless. This is the only place where we touch `dhcp_helper_procs`, , and there should never be more than one instance of `IPAllocatorDHCP`.

nstng commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -435,30 +439,22 @@ def _release_dhcp_ip(self, ip_desc: IPDesc) -> None:
+            if len(self.dhcp_helper_procs) == self.MAX_DHCP_PROCS:
+                oldest_proc = self.dhcp_helper_procs.pop()
+                if oldest_proc.poll() is None:
+                    oldest_proc.wait(timeout=10)
```
Iiuc this can throw a TimeoutExpired exception and by this prevent the current call from opening a subprocess.

MoritzThomasHuebner replied:
I added error handling for `TimeoutExpired`. If the process does not finish within 10 seconds, we kill it and move on.  However, there is still a problem: If we kill the release process and the release did not succeed, we will still have deleted the entry in the MobilityStore. I am not sure how we can mitigate this issue.

nstng replied:
Independent on how the remaining issue plays out - I suggest to refactor the complete while-part into an additional function.

> If we kill the release process and the release did not succeed, we will still have deleted the entry in the MobilityStore. I am not sure how we can mitigate this issue.

Iiuc then this happens before the `_release_dhcp_ip` call, i.e., we always had the potential that the store and the dhpc (is this only redis or also the dhpc server itself?) are out of sync. The only real mitigation I see here is a regular in-sync check with clean up - but imho this is out of scope here. Do we know of implications when store and redis are out of sync?

MoritzThomasHuebner replied:
I added your requested changes in the latest commit.

MoritzThomasHuebner replied:
The changes we discussed offline

nstng reviewed (changes requested):
The tests are green for me vs a debian based environment. For changes requested see comments. At least the error handling should be investigated.

MoritzThomasHuebner committed:
feat(agw): Improved error handling

- Improved thread safety by ensuring that no more than 3 release processes can be run at the same time by using '>=' instead of '==' in a check.
- Added error handling for stuck release processes.

Co-authored-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Extract while loop for deque step

Co-authored-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Add udhcpd traffic server state check

- Added check to verify state of leases during non nat tests
    - Count to check that the correct number of leases are assigned during attach
    - Verify that all leases expire after detach
- Modified `exec_command` in traffic_util.py to return a `CompletedProcess`
    - Adapted existing calls to `exec_command` so that they only look at the return code
- Added new `TrafficUtil` methods `dump_leases` and `clear_leases`
    - `dump_leases` returns the stdout of running `dumpleases` on the traffic server
    - `clear_leases` removes all existing leases from the traffic server (used during `tearDown`, the new checks wouldn't work otherwise).

Co-authored-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +112,65 @@ def __init__(self):
 
+    def dump_leases(self):
+        """Dump DHCP leases in TRF server VM"""
+        return self.exec_command("dumpleases").stdout
```
Why is `capture_output=True` not needed here?

MoritzThomasHuebner replied:
Good point, I believe I missed running make again before re-running the tests.

MoritzThomasHuebner committed:
refactor(agw): Extracted and improved lease checks

Co-authored-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp_multi_ue.py:
```diff
@@ -0,0 +1,94 @@
+        self.trf_util.clear_leases()
+        self._s1ap_wrapper.cleanup()
+
+    def test_attach_detach_with_non_nat_dhcp(self):
```
Please rename to test_attach_detach_with_non_nat_dhcp_multi_ue - undocumented convention.

MoritzThomasHuebner replied:
Done

nstng commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp_multi_ue_looped.py:
```diff
@@ -0,0 +1,93 @@
+        self.trf_util.clear_leases()
+        self._s1ap_wrapper.cleanup()
+
+    def test_attach_detach_with_non_nat_dhcp(self):
```
Please rename to test_attach_detach_with_non_nat_dhcp_multi_ue_looped - undocumented convention.

MoritzThomasHuebner replied:
Done

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+            capture_output=capture_output,
         )
 
+    def dump_leases(self):
```
NIT: private function

MoritzThomasHuebner replied:
fixed

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+        cmd = 'systemctl stop udhcpd.service && ' \
+              'rm -f /var/lib/misc/udhcpd.leases && ' \
+              'systemctl start udhcpd.service'
+        return self.exec_command(f"sudo bash -c '{cmd}'").returncode
```
NIT: return value never used

MoritzThomasHuebner replied:
I removed the return statement

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+        return self.exec_command(f"sudo bash -c '{cmd}'").returncode
+
+    @property
+    def n_leases(self):
```
NIT: can be private, name: why not count_leases?

MoritzThomasHuebner replied:
I used a "n_leases", i.e. a noun because I made this a property so access would to it would be the same as an attribute. I think it would be alright, but I refactored it to just be a regular method `_count_leases`.

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+    @property
+    def n_leases(self):
+        """Count the total number of leases in TRF server VM"""
+        return len(self.dump_leases().decode("utf-8").split("\n")) - 2
```
NIT: maybe a small comment why the magic `2` is subtracted? I assume some standard output?

MoritzThomasHuebner replied:
Done

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+        return len(self.dump_leases().decode("utf-8").split("\n")) - 2
+
+    @property
+    def n_expired_leases(self):
```
NIT: can be private, name: why not count_expired_leases?

MoritzThomasHuebner replied:
See my comment regarding n_leases

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+        """Count number of expired leases in TRF server VM"""
+        return self.dump_leases().decode("utf-8").count("expired")
+
+    def check_attached_leases(self, expected_leases, timeout=5):
```
NIT: the timeout is never changed in a call - why not hard code it for now and have it changed if at some point someone needs another timeout?

MoritzThomasHuebner replied:
Done

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+            time.sleep(1)
+            i += 1
+
+    def check_detached_leases(self, expected_leases, timeout=5):
```
NIT: timeout parameter as above

MoritzThomasHuebner replied:
Done

nstng commented on lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py:
```diff
@@ -112,55 +113,93 @@ def __init__(self):
+    def check_detached_leases(self, expected_leases, timeout=5):
+        """Wait for up to timeout seconds for expected number of leases"""
+        i = 0
+        while expected_leases != self.n_expired_leases:
```
In my tests the loop was never entered - did you had cases where it took some time until all leases were expired? I think the loop/wait is still reasonable - just wondering.

MoritzThomasHuebner replied:
I believe this happened once for me. It would be a relatively rare case, but the leases are first removed in redis and then on the traffic server. It would not be out of sync for a long time, but in the test code the release loop might could exactly end when the last redis entry is removed but the traffic server is still out of sync. You might be able to trigger that case more often by removing the 5 second sleep from the release loop in the test code.

nstng commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -435,36 +439,39 @@ def _release_dhcp_ip(self, ip_desc: IPDesc) -> None:
+                        "Killing release process "
+                        "and moving on.", ip_desc.ip,
+                    )
+                    oldest_proc.kill()
```
Iirc we talked about whether `.kill()` could rise an exception (an exception would prevent the follow up call to the helper CLI). Do you have an answer to that?

MoritzThomasHuebner replied:
Killing/terminating an already expired process does not cause an exception.
```
Python 3.8.10 (default, Jul 12 2022, 09:44:48) 
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> proc = subprocess.Popen(["echo", "0"])
>>> proc.wait()
0
>>> proc.returncode
0
>>> proc.kill()
>>> proc
<subprocess.Popen object at 0x7f8d4235ccd0>
```

MoritzThomasHuebner committed:
refactor(agw): Renamed test functions

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Some more refactoring

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Removed a return statement

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

nstng reviewed (approved):
Thank you for applying all the review comments :+1: lgtm
